packages/typespec-azure-resource-manager/src/common-types.ts

```diff
@@ -56,11 +55,19 @@ export function isArmCommonType(entity: Type): boolean {
 export function $armCommonTypesVersion(
   context: DecoratorContext,
   entity: Namespace | EnumMember,
-  version: string | EnumValue
+  version: EnumMember
 ) {
-  context.program
-    .stateMap(ArmStateKeys.armCommonTypesVersion)
-    .set(entity, typeof version === "string" ? version : version.value.name);
+  context.program.stateMap(ArmStateKeys.armCommonTypesVersion).set(entity, version.name as string);
+
+  if (entity.kind === "Namespace") {
+    const versioned = entity.decorators.find((x) => x.definition?.name === "@versioned");
+    // If it is versioned namespace, we will skip adding @useDependency to namespace
+    if (versioned) {
+      return;
```

Comment:
This doesn't work in all cases, for example: when I add common-types version to a versioned spec, I can add it once to the namespace for all versions [like this](https://azure.github.io/typespec-azure/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZcmVzdNUZdmVyc2lvbmluZ8wfYXp1cmUtdG9vbHMvyCstxhVjb3Jl3yvIK3Jlc291cmNlLW1hbmFnZXIiOwoKdXNpbmcgVHlwZVNwZWMuSHR0cDvQFVJlc3TRFVbpAKnIG0HEZS5Db3JlzhJSx3dNxnY7CgovKiogQ29udG9zbyDIHiBQcm92aWRlciDmAJ5tZW50IEFQSS4gKi8KQGFybcggTmFtZXNwYWNlCkBzZXJ2aWNlKHsKICB0aXRsZTogIsdYyC5IdWJDbGllbnQiLAp9KQpA5wFgZWQo5wDBcynFXkNvbW1vbuQA3nPHGSj1AL4uyynoAQRzLnY1KQpu6ACYIE1pY3Jvc29mdC7yAI3vAP5BUEkg5wCTc%2BQA6WVudW0gyFcg5ADYxC4yMDIxLTEwLTAxLXByZXZpZXfINcQ0ICBAdXNlRGVwZW5kZW5jefcAuchWLnYxXzBfUMZIXzEpCiAgYNJhYCwK333ffd99zn0y0X195gEoQegBKusBQyDoAsjkAIBtb2RlbCBFbXBsb3llZSBpcyBUcmFja2Vk6ACCPMgcUHJvcGVydGllcz7lAVkuLukApuQCU1BhcmFtZXRlcskxPjvoAIbJX3DJRNJ8ymDpAbhBZ2Ugb2YgZcg%2F5QEwYWdlPzogaW50MzI76AFkQ2l0edIqY2l0eT86IHN0cuUDZscsUHJvZmls01lAZW5jb2RlKCJiYXNlNjR1cmwi5AFgcMYwPzogYnl0ZXPJSFRoZSBzdGF0dXPES3RoZSBsYXN0IOQAxWF0aW9u5QN5ICBAdmlzaWJpbGl0eSgicmVhZMhVxBjlBABTdGF0xF%2FlAaPMFOkBQMRrzDLFeOUAwmHpAc3Fb0Bscm%2FEO3VzCnVu5AJi0VTlAVzmARLpAqLEX8hHIGNyZcQncmVxdWVzdCBoYXMgYmVlbiBhY2NlcHRlZMRnICBBxw46ICLICyLWUGnEQOQAtOkAwchE7ACcOiAizA%2FaTHVwZGF0xE%2FFQ1XHDjogIsgLyjvpBRrpAMTmANxk5wGaU3VjY2VlZOUAxckM0z%2FFNuQBTWZhaWzJPkbFDTogIsYJ3Dh3YXMgY2FuY2XKPkPHD%2BQFcccL%2FwFAIGRlbGXpAYBExA3mAPnIC%2BQFoe8Db21vduoByukDcU1vdmVSxxXoA2rEc21vdsRoZnJvbSBsb2PmALzFbsQT8QNIyzN0b88xdG%2FKL%2FcAlXNwb25z6wSD5gCWxxbsAJfuA2TFPsVkxnzmAu7NbmludGVyZuQGSU%2FoA4xzIGV4dGVuZHP2B1Muyyl7fQrlBtHII8oby1noAM3mBKtnZXTkAadByjVSZWFk7AUIIOcCe09y5QKn5QHWyy9Dxx1SZXBsYWNlQXN5bmPOP%2BUC99A3UGF0Y2hTzCws8wVOxUDmAj3PQOUCOmVXaXRob3V0T2vTd2xpc3RCecgwR3JvdXDPREzFIlBhcmVudNQ8U3Vic2NyaXDlAhzGO8YzzBnMOegF%2ByBzYW1wbOsDCWFjxUR0aGF05gIG6QW%2BdG8gZGlmZmXkAITvAonFKe4AskHFSO8BN%2BsDFMgN5gKL8wCSSEVBROoF58R%2BY2hlY2vqAKpleGlzdGVu5gdhIMYeRckU7wH7zR3uBzw%3D&e=%40azure-tools%2Ftypespec-autorest&options=%7B%22linterRuleSet%22%3A%7B%22extends%22%3A%5B%22%40azure-tools%2Ftypespec-azure-rulesets%2Fresource-manager%22%5D%7D%7D)

Reply:
Logic has been updated.

from @armProviderNamespace:
    - look up current @armCommonTypeVersion on namespace if any
    - Versioned, loop thru each service version enum member
    -       if no @armCommonTypeVersion, add and set version to either the one from step above or default to v3
    - Unversioned, add and set version to either the one from step above or default to v3 to namespace if one doesn't exist